src/EditorFeatures/Core/Shared/Tagging/TagSources/ProducerPopulatedTagSource.cs

```diff
@@ -221,10 +221,46 @@ private void OnChanged(object sender, TaggerEventArgs e)
 
         private void AccumulateTextChanges(TextContentChangedEventArgs contentChanged)
         {
-            var textChangeRanges = contentChanged.Changes.Select(c => new TextChangeRange(new TextSpan(c.OldSpan.Start, c.OldSpan.Length), c.NewLength)).ToList();
-            lock (_cachedTagsGate)
+            var contentChanges = contentChanged.Changes;
+            var count = contentChanges.Count;
+
+            switch (count)
             {
-                _accumulatedTextChanges = _accumulatedTextChanges.Accumulate(textChangeRanges);
+                case 0:
+                    return;
+
+                case 1:
+                    // PERF: Optimize for the simple case of typing on a line.
+                    {
+                        var c = contentChanges[0];
+                        var textChangeRange = new TextChangeRange(new TextSpan(c.OldSpan.Start, c.OldSpan.Length), c.NewLength);
+                        lock (_cachedTagsGate)
+                        {
+                            if (_accumulatedTextChanges == null)
+                            {
+                                _accumulatedTextChanges = textChangeRange;
+                            }
+                            else
+                            {
+                                _accumulatedTextChanges = _accumulatedTextChanges.Accumulate(SpecializedCollections.SingletonEnumerable(textChangeRange));
```

Comment:
SingletonEnumerable allocates, so I'm trying to avoid it in the null case because that's the common case (when there are no accumulated changes). Doing the interlocked loop thing will be pretty delicate and, as I mentioned previously, the lock is used to protect other fields in the class so I don't want to risk getting things out of step (letting the interlocked succeed while another thread holds the lock to update the  'cached tag' fields)